tests/cpp/aot/aot_save_load_test.cpp

```diff
@@ -91,19 +95,63 @@ TEST(AotSaveLoad, Vulkan) {
 
   aot_save();
 
-  vulkan::AotModuleLoaderImpl aot_loader(".");
-  vulkan::VkRuntime::RegisterParams init_kernel, ret_kernel;
+  // API based on proposal https://github.com/taichi-dev/taichi/issues/3642
+  // Initialize Vulkan program
+  taichi::uint64 *result_buffer{nullptr};
+  taichi::lang::RuntimeContext host_ctx;
+  auto memory_pool =
+      std::make_unique<taichi::lang::MemoryPool>(Arch::vulkan, nullptr);
+  result_buffer = (taichi::uint64 *)memory_pool->allocate(
+      sizeof(taichi::uint64) * taichi_result_buffer_entries, 8);
+
+  // Create Taichi Device for computation
+  lang::vulkan::VulkanDeviceCreator::Params evd_params;
+  evd_params.api_version =
+      taichi::lang::vulkan::VulkanEnvSettings::kApiVersion();
+  auto embedded_device =
+      std::make_unique<taichi::lang::vulkan::VulkanDeviceCreator>(evd_params);
+
+  // Create Vulkan runtime
+  vulkan::VkRuntime::Params params;
+  params.host_result_buffer = result_buffer;
+  params.device = embedded_device->device();
+  auto vulkan_runtime =
+      std::make_unique<taichi::lang::vulkan::VkRuntime>(std::move(params));
+
+  // Run AOT module loader
+  vulkan::AotModuleParams mod_params;
+  mod_params.module_path = ".";
+  mod_params.runtime = vulkan_runtime.get();
+
+  std::unique_ptr<aot::Module> vk_module =
+      aot::Module::load(".", Arch::vulkan, mod_params);
+  EXPECT_TRUE(vk_module);
+
+  // Retrieve kernels/fields/etc from AOT module to initialize runtime
+  auto root_size = vk_module->get_root_size();
+  EXPECT_EQ(root_size, 64);
+  vulkan_runtime->add_root_buffer(root_size);
 
-  auto ret = aot_loader.get_kernel("init", init_kernel);
-  EXPECT_TRUE(ret);
+  // TODO
+  // auto init_kernel = vk_module->get_kernel("init");
+  // EXPECT_TRUE(init_kernel);
 
-  ret = aot_loader.get_kernel("ret", ret_kernel);
-  EXPECT_TRUE(ret);
+  // auto substep_kernel = vk_module->get_kernel("substep");
```

Comment:
Why not fix this?

Reply:
i can fix these all in this PR if that is better

Reply:
Yep, please do.